### PR TITLE
Fix precision error in RenderSliverFixedExtentBoxAdaptor assertion

### LINKED ISSUE #95101: Precision error in RenderSliverFixedExtentBoxAdaptor
https://github.com/flutter/flutter/pull/30979 fixed the index detection error, but it didn't consider that doing so in its turn also creates another precision error on this line.

https://github.com/flutter/flutter/blob/a3dc90c4f51523fd3f8df6eee3d79239a3c8de52/packages/flutter/lib/src/rendering/sliver_fixed_extent_list.dart#L270

Consider an example where:
```
scrollOffset = 392.7272727256901
itemExtent = 392.72727272727275
```

In this case `getMinChildIndexForScrollOffset` already returns 1 because of the `scrollOffset / itemExtent` and the precision error is taken from this.

But later the assertion in question checks for precision of  `childScrollOffset(firstChild!)! - itemExtent`, where `childScrollOffset` (I assume in this case in particular, not always, but I'm not sure) returns `itemExtent`, which gives us a value bigger than `precisionErrorTolerance` and fails the assertion.

## Steps to Reproduce

1. Execute `flutter test` on the code sample

**Expected results:** Test passes

**Actual results:** Test fails

<details>
<summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  testWidgets('PageView rounding issue', (WidgetTester tester) async {
    final PageController controller = PageController(initialPage: 152);

    await tester.pumpWidget(
      Center(
        child: SizedBox(
          width: 392.72727272727275,
          child: Directionality(
            textDirection: TextDirection.ltr,
            child: PageView.builder(
              controller: controller,
              itemCount: 366,
              itemBuilder: (BuildContext context, int index) {
                return const SizedBox();
              },
            ),
          ),
        ),
      ),
    );

    controller.jumpToPage(365);
    await tester.pump();
  });
}
```
</details>

<details>
  <summary>Logs</summary>

```
══╡ EXCEPTION CAUGHT BY RENDERING LIBRARY ╞═════════════════════════════════════════════════════════
The following assertion was thrown during performLayout():
'package:flutter/src/rendering/sliver_fixed_extent_list.dart': Failed assertion: line 270 pos 12:
'firstIndex == 0 || childScrollOffset(firstChild!)! - scrollOffset <= precisionErrorTolerance': is
not true.

Either the assertion indicates an error in the framework itself, or we should provide substantially
more information in this error message to help you determine and fix the underlying cause.
In either case, please report this assertion by filing a bug on GitHub:
  https://github.com/flutter/flutter/issues/new?template=2_bug.md

When the exception was thrown, this was the stack:
#2      RenderSliverFixedExtentBoxAdaptor.performLayout (package:flutter/src/rendering/sliver_fixed_extent_list.dart:270:12)
#3      RenderObject.layout (package:flutter/src/rendering/object.dart:1882:7)
#4      RenderSliverEdgeInsetsPadding.performLayout (package:flutter/src/rendering/sliver_padding.dart:137:12)
#5      _RenderSliverFractionalPadding.performLayout (package:flutter/src/widgets/sliver_fill.dart:167:11)
#6      RenderObject.layout (package:flutter/src/rendering/object.dart:1882:7)
#7      RenderViewportBase.layoutChildSequence (package:flutter/src/rendering/viewport.dart:510:13)
#8      RenderViewport._attemptLayout (package:flutter/src/rendering/viewport.dart:1580:12)
#9      RenderViewport.performLayout (package:flutter/src/rendering/viewport.dart:1489:20)
#10     RenderObject._layoutWithoutResize (package:flutter/src/rendering/object.dart:1729:7)
#11     PipelineOwner.flushLayout (package:flutter/src/rendering/object.dart:887:18)
#12     AutomatedTestWidgetsFlutterBinding.drawFrame (package:flutter_test/src/binding.dart:1131:23)
#13     RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:363:5)
#14     SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1145:15)
#15     SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:1082:9)
#16     AutomatedTestWidgetsFlutterBinding.pump.<anonymous closure> (package:flutter_test/src/binding.dart:995:9)
#19     TestAsyncUtils.guard (package:flutter_test/src/test_async_utils.dart:71:41)
#20     AutomatedTestWidgetsFlutterBinding.pump (package:flutter_test/src/binding.dart:982:27)
#21     WidgetTester.pump.<anonymous closure> (package:flutter_test/src/widget_tester.dart:608:53)
#24     TestAsyncUtils.guard (package:flutter_test/src/test_async_utils.dart:71:41)
#25     WidgetTester.pump (package:flutter_test/src/widget_tester.dart:608:27)
#26     main.<anonymous closure> (file:///C:/dev/src/flutter/packages/flutter/test/widgets/page_view_test.dart:1123:18)
<asynchronous suspension>
<asynchronous suspension>
(elided 7 frames from class _AssertionError, dart:async, and package:stack_trace)

The following RenderObject was being processed when the exception was fired: RenderSliverFillViewport#4cfb6 relayoutBoundary=up2:
  needs compositing
  creator: _SliverFillViewportRenderObjectWidget ← _SliverFractionalPadding ← SliverFillViewport ←
    Viewport ← IgnorePointer-[GlobalKey#e98c5] ← Semantics ← Listener ← _GestureSemantics ←
    RawGestureDetector-[LabeledGlobalKey<RawGestureDetectorState>#78667] ← Listener ← _ScrollableScope
    ← _ScrollSemantics-[GlobalKey#80202] ← ⋯
  parentData: paintOffset=Offset(0.0, 0.0) (can use size)
  constraints: SliverConstraints(AxisDirection.right, GrowthDirection.forward, ScrollDirection.idle,
    scrollOffset: 143345.5, remainingPaintExtent: 392.7, crossAxisExtent: 600.0, crossAxisDirection:
    AxisDirection.down, viewportMainAxisExtent: 392.7, remainingCacheExtent: 392.7, cacheOrigin: 0.0)
  geometry: SliverGeometry(scrollExtent: 143738.2, paintExtent: 392.7, maxPaintExtent: 143738.2,
    hasVisualOverflow: true, cacheExtent: 392.7)
  currently live children: 365 to 365
This RenderObject had the following descendants (showing up to depth 5):
    child with index 365: RenderIndexedSemantics#922dc
      child: RenderRepaintBoundary#e581e
        child: RenderConstrainedBox#3f624
════════════════════════════════════════════════════════════════════════════════════════════════════
Test failed. See exception logs above.
The test description was: PageView rounding issue

✖ PageView rounding issue
Exited (1)
```

```
[√] Flutter (Channel unknown, 2.6.0-12.0.pre.999, on Microsoft Windows [Version 10.0.19041.1348], locale ru-RU)
    • Flutter version 2.6.0-12.0.pre.999 at C:\dev\src\flutter
    • Upstream repository unknown
    • FLUTTER_GIT_URL = git@github.com:nt4f04und/flutter.git
    • Framework revision 895beb04bb (23 hours ago), 2021-12-10 18:54:10 -0800
    • Engine revision 79f750d4a5
    • Dart version 2.16.0 (build 2.16.0-85.0.dev)
    • DevTools version 2.9.1

[!] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at C:\Users\danya\AppData\Local\Android\sdk
    X cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    X Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/windows#android-setup for more details.
[√] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop for Windows (Visual Studio Community 2019 16.7.7)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.7.30621.155
    • Windows 10 SDK version 10.0.19041.0

[√] Android Studio (version 4.1)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)

[√] IntelliJ IDEA Community Edition (version 2020.3)
    • IntelliJ at C:\Program Files\JetBrains\IntelliJ IDEA Community Edition 2020.3.3
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart

[√] VS Code (version 1.63.0)
    • VS Code at C:\Users\danya\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.29.0

[√] Connected device (4 available)
    • sdk gphone x86 (mobile) • emulator-5554 • android-x86    • Android 11 (API 30) (emulator)
    • Windows (desktop)       • windows       • windows-x64    • Microsoft Windows [Version 10.0.19041.1348]
    • Chrome (web)            • chrome        • web-javascript • Google Chrome 96.0.4664.93
    • Edge (web)              • edge          • web-javascript • Microsoft Edge 96.0.1054.34

! Doctor found issues in 1 category
```
</details>

### PR DESCRIPTION
**Pull request by nt4f04uNd**
Fixes https://github.com/flutter/flutter/issues/95101

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/rendering/sliver_fixed_extent_list.dart [hunk 1]
```diff
@@ -267,7 +267,10 @@ abstract class RenderSliverFixedExtentBoxAdaptor extends RenderSliverMultiBoxAda
     final double leadingScrollOffset = indexToLayoutOffset(itemExtent, firstIndex);
     final double trailingScrollOffset = indexToLayoutOffset(itemExtent, lastIndex + 1);
 
-    assert(firstIndex == 0 || childScrollOffset(firstChild!)! - scrollOffset <= precisionErrorTolerance);
+    // Allow a slightly bigger precision error, because firstIndex already might have
+    // additional precision error, which comes from `getMinChildIndexForScrollOffset`,
+    // which also uses `precisionErrorTolerance`.
+    assert(firstIndex == 0 || childScrollOffset(firstChild!)! - scrollOffset <= precisionErrorTolerance * 10);
```
Comment (chunhtai) [inline packages/flutter/lib/src/rendering/sliver_fixed_extent_list.dart:273]: It is weird we multiply the precision error by a constant scaling.

I would do it in the getMinChildIndexForScrollOffset

specifically, by changine the if case to the following
```
      if ((actual * itemExtent - round * itemExtent).abs() < precisionErrorTolerance) {
        return round;
      }
```

Comment (nt4f04uNd) [reply packages/flutter/lib/src/rendering/sliver_fixed_extent_list.dart]: Ah, ok. I changed `getMaxChildIndexForScrollOffset` aswell for consistency.

File: packages/flutter/test/widgets/page_view_test.dart [hunk 2]
```diff
@@ -1096,4 +1096,30 @@ void main() {
...
+                return const SizedBox();
+              },
+            ),
+          ),
+        ),
+      ),
+    );
+
+    controller.jumpToPage(365);
+    await tester.pump();
```
Comment (chunhtai) [inline packages/flutter/test/widgets/page_view_test.dart:1123]: The test should expect for something

maybe add an `expect(tester.takeException(), isNull)`

**Review by chunhtai — APPROVED** [review]
LGTM

**Comment by nt4f04uNd** [comment]
Actually, failed tests indicate that precision gets worse from this. I debugged it and it turns out, the bigger the item extent is, the bigger precision loss.

In the code below if we increase `i` the printed value gets bigger and bigger

```dart
void main() {
  final n = 10 + 1.666666804567285e-11;
  final r = n.round();
  const i = 1000;
  print(n * i - r * i);
}

```

**Comment by chunhtai** [comment]
> Actually, failed tests indicate that precision gets worse from this. I debugged it and it turns out, the bigger the item extent is, the bigger precision loss.
> 
> In the code below if we increase `i` the printed value gets bigger and bigger
> 
> ```dart
> void main() {
>   final n = 10 + 1.666666804567285e-11;
>   final r = n.round();
>   const i = 1000;
>   print(n * i - r * i);
> }
> ```

yes we compares bigger value against the precisionErrorTolerance, that means the check will fail easier. and goes to the floor case, or did I misunderstand something?

**Comment by nt4f04uNd** [comment]
> yes we compare bigger value against the precisionErrorTolerance

That is the problem, isn't it? It should not increase - we lose precision where we otherwise would not.

**Comment by chunhtai** [comment]
> > yes we compare bigger value against the precisionErrorTolerance
> 
> That is the problem, isn't it? It should not increase - we lose precision where we otherwise would not.

I think the correct comparison should be something like this

```
  final error = 1.666666804567285e-11;
  final scrollOffset = 100000 + error;
  const itemExtent = 10000;
  double actual = scrollOffset/itemExtent;
  final r = actual.round();
  print(actual * itemExtent - r * itemExtent);
```
It only make sense if scrollOffset is (multiple of itemExtent)  + error. The goal is to make sure the error does not change too much, neither smaller nor bigger.

**Comment by nt4f04uNd** [comment]
Ah, sorry for the confusion! Please take a look

**Review by chunhtai — APPROVED** [review]
LGTM, thanks for the fix!

**Pull request merged**